airflow/datasets/manager.py

```diff
@@ -44,6 +45,13 @@ class DatasetManager(LoggingMixin):
     def __init__(self, **kwargs):
         super().__init__(**kwargs)
 
+    def create_datasets(self, dataset_models: list[DatasetModel], session: Session) -> None:
+        """Create new datasets."""
+        for dataset_model in dataset_models:
+            session.add(dataset_model)
+            self.notify_dataset_created(dataset=Dataset(uri=dataset_model.uri, extra=dataset_model.extra))
+        session.flush()  # this is required to ensure each dataset has its PK loaded
```

Comment:
This called the hook before the database is flushed (i.e. before the rows can actually appear in the database), while the changed hook is called after flushing. I feel this should be more consistent.

Reply:
Addressed in https://github.com/apache/airflow/pull/34418/commits/c471bde88ba025d098c541b35e38d880ce368776
`on_dataset_created` hook is now called after the database is flushed.